test/metabase/notification/send_test.clj

```diff
@@ -184,3 +187,61 @@
               (is (prometheus-test/approx= 1 (mt/metric-value system :metabase-notification/channel-send-error
                                                               {:payload-type "notification/testing"
                                                                :channel-type "channel/metabase-test"}))))))))))
+
+(deftest notification-dispatcher-test
+  (testing "notification dispatcher"
+    (let [sent-notifications  (atom [])
+          wait-for-processing #(u/poll {:thunk       (fn [] (count @sent-notifications))
+                                        :done?       (fn [cnt] (= cnt %))
+                                        :interval-ms 10
+                                        :timeout-ms  1000})]
+      (with-redefs [notification.send/send-notification-sync! (fn [notification]
+                                                                ;; fake latency
+                                                                (Thread/sleep 20)
+                                                                (swap! sent-notifications conj notification))]
+        (let [test-dispatcher (#'notification.send/create-notification-dispatcher 2)]
+          (testing "basic processing"
+            (reset! sent-notifications [])
+            (let [notification {:id 1 :test-value "A"}]
+              (test-dispatcher notification)
+              (wait-for-processing 1)
+              (is (= [notification] @sent-notifications))))
+
+          (testing "notifications without IDs are all processed"
+            (reset! sent-notifications [])
+            (test-dispatcher {:test-value "B"})
+            (test-dispatcher {:test-value "C"})
+            (wait-for-processing 2)
+            (is (= 2 (count @sent-notifications)))
+            (is (= #{"B" "C"} (into #{} (map :test-value @sent-notifications)))))
+
+          (testing "notifications with same ID are replaced in queue"
+            (reset! sent-notifications [])
+            ;; make the queue busy
+            (test-dispatcher {:id 40 :test-value "D"})
+            (test-dispatcher {:id 41 :test-value "D"})
+            (test-dispatcher {:id 42 :test-value "D"})
+            (test-dispatcher {:id 42 :test-value "E"})
+            (u/poll {:thunk       (fn [] (->> @sent-notifications
+                                              (filter #(= 42 (:id %)))
+                                              first :test-value))
+                     :done?       (fn [value] (= "E" value))
```

Comment:
OK, so this is why you have that `id->notification` map. I thought we'd want to get "D".

Can you lay out what properties of the notification are such that we want "keep last" semantics?

Reply:
so when submitting a notification you can actually send a fully hydrated notification( payload info, handlers etc)

I'd say the most important ones are payload info and handlers.

Reply:
You answered me on the main thread, so I'll resolve this 😄